service/dap/server_test.go

```diff
@@ -4433,11 +4433,6 @@ func TestOptionalNotYetImplementedResponses(t *testing.T) {
 		client.RestartRequest()
 		expectNotYetImplemented("restart")
 
-		client.StepBackRequest()
```

Comment:
Is it possible to add tests for these requests?

Reply:
Let me familiarize with the test setups and I can look into it

Reply:
Something I'm noticing (please correct me if I'm wrong) is that I'd have to add support for debug sessions without program fxtures to have something like this:

```go
func TestLaunchRequestNoProgram(t *testing.T) {
	runNoProgramTest(t, func(client *daptest.Client) {
		runNoProgramDebugSession(t, client, func() {
			client.LaunchRequestWithArgs(map[string]interface{}{
				"mode": "core", "coreFilePath": "./_fixtures/trace/core1"})
		}, 0)
	})
}
```

And I still have two questions about this code (it's a WIP I have on macOS but that I'll have to test on linux of course):

- The tests will be have to be skipped on unsupported OS (windows for core mode and replay is linux only ATM)
- I have to solve the fixture loading from the trace files in a similar way the protest package does it for the test names

Reply:
The later seems to be the blocker right now for me:

```
/usr/local/go/bin/go tool test2json -t /private/var/folders/b5/j00pg5rx2y924slkhhjlhgpjcp402s/T/___TestLaunchRequestNoProgram_in_github_com_go_delve_delve_service_dap.test -test.v -test.run ^\QTestLaunchRequestNoProgram\E$
=== RUN   TestLaunchRequestNoProgram
DAP server listening at: [::]:61608
Connecting to server at: [::]:61608
    server_test.go:3898: got &dap.ErrorResponse{Response:dap.Response{ProtocolMessage:dap.ProtocolMessage{Seq:0, Type:"response"}, RequestSeq:2, Success:false, Command:"launch", Message:"Failed to launch"}, Body:dap.ErrorResponseBody{Error:dap.ErrorMessage{Id:3000, Format:"Failed to launch: open foo/bar: no such file or directory", Variables:map[string]string(nil), SendTelemetry:false, ShowUser:false, Url:"", UrlLabel:""}}}, want *dap.LaunchResponse
--- FAIL: TestLaunchRequestNoProgram (0.10s)

FAIL

Process finished with the exit code 1
```

As I mentioned, this must work on linux _and_ windows so it's going to be tricky

Reply:
As far as I am concerned the code is straightforward enough that it doesn't need tests. Also any tests added to this will not ever run in CI because nothing we have supports rr.

Reply:
I do see some "rr" related tests in the codebase (that run with `protest.MustHaveRecordingAllowed(t)` and `protest.AllowRecording(t)`). Do they only run locally? What about coredump testing?

If getting an end-to-end replay/core test is too tricky, there are still other useful things we can test that I mentioned earlier, but did not get a response for. For example, there have been questions and revisions in this PR about valid/invalid combinations of launch attributes, so l think it would be useful to add bad coverage to TestBadLaunchRequests to "document" this contract.

Reply:
> Do they only run locally?

Yes. And only if you have rr installed.

> What about coredump testing?

See withCoreFile in `pkg/proc/core/core_test.go` and TestMinidump in the same file. You would have to replicate them in server_test.go, seems like a questionable way to spend time to me.

Reply:
I see. Yeah, then it's not worth it. Lets only cover launch attributes handling then.